whitespace.go

```diff
@@ -0,0 +1,283 @@
+			Run(p, settings)
+			return nil, nil
+		},
+		RunDespiteErrors: true,
+	}
+}
+
+func flags(settings *Settings) flag.FlagSet {
+	flags := flag.NewFlagSet("", flag.ExitOnError)
+	flags.BoolVar(&settings.MultiIf, "multi-if", settings.MultiIf, "Check that multi line if-statements have a leading newline")
+	flags.BoolVar(&settings.MultiFunc, "multi-func", settings.MultiFunc, "Check that multi line functions have a leading newline")
+
+	return *flags
+}
+
+func Run(pass *analysis.Pass, settings *Settings) []Message {
+	messages := []Message{}
+
+	for _, file := range pass.Files {
+		filename := pass.Fset.Position(file.Pos()).Filename
+		if !strings.HasSuffix(filename, ".go") {
+			continue
+		}
+
+		fileMessages := runFile(file, pass.Fset, *settings)
+
+		if settings.Mode == RunningModeGolangCI {
+			messages = append(messages, fileMessages...)
+			continue
+		}
+
+		for _, message := range fileMessages {
+			pass.Report(analysis.Diagnostic{
+				Pos:      message.Diagnostic,
+				Category: "whitespace",
+				Message:  message.Message,
+				SuggestedFixes: []analysis.SuggestedFix{
+					{
+						TextEdits: []analysis.TextEdit{
+							{
+								Pos:     message.FixStart,
+								End:     message.FixEnd,
+								NewText: []byte("\n"),
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+
+	return messages
+}
+
+func runFile(file *ast.File, fset *token.FileSet, settings Settings) []Message {
+	var messages []Message
+
+	for _, f := range file.Decls {
+		decl, ok := f.(*ast.FuncDecl)
+		if !ok || decl.Body == nil { // decl.Body can be nil for e.g. cgo
+			continue
+		}
+
+		vis := visitor{file.Comments, fset, nil, make(map[*ast.BlockStmt]bool), settings}
+		ast.Walk(&vis, decl)
+
+		messages = append(messages, vis.messages...)
+	}
+
+	return messages
+}
+
+type visitor struct {
+	comments    []*ast.CommentGroup
+	fset        *token.FileSet
+	messages    []Message
+	wantNewline map[*ast.BlockStmt]bool
+	settings    Settings
+}
+
+func (v *visitor) Visit(node ast.Node) ast.Visitor {
+	if node == nil {
+		return v
+	}
+
+	if stmt, ok := node.(*ast.IfStmt); ok && v.settings.MultiIf {
+		checkMultiLine(v, stmt.Body, stmt.Cond)
+	}
+
+	if stmt, ok := node.(*ast.FuncLit); ok && v.settings.MultiFunc {
+		checkMultiLine(v, stmt.Body, stmt.Type)
+	}
+
+	if stmt, ok := node.(*ast.FuncDecl); ok && v.settings.MultiFunc {
+		checkMultiLine(v, stmt.Body, stmt.Type)
+	}
+
+	if stmt, ok := node.(*ast.BlockStmt); ok {
+		wantNewline := v.wantNewline[stmt]
+
+		comments := v.comments
+		if wantNewline {
+			comments = nil // Comments also count as a newline if we want a newline
+		}
+
+		first, last := firstAndLast(comments, v.fset, stmt.Pos(), stmt.End(), stmt.List)
+		startMsg := checkStart(v.fset, stmt.Lbrace, first)
+
+		if wantNewline && startMsg == nil && len(stmt.List) >= 1 {
+			v.messages = append(v.messages, Message{
+				Diagnostic:  stmt.Lbrace,
+				FixStart:    stmt.List[0].Pos(),
+				FixEnd:      stmt.List[0].Pos(),
```

Comment:
It this correct? Or should this be `.End()`?

Reply:
Yeah it's a bit weird maybe but why I added the split between where the diagnostic is and where the fix is. Basically this just says that from where the first statement starts, add a newline. So it's not replacing anything because the range is zero, it's just adding.

```go
func fn() {
    stmt.List[0]
   ^
    |__From here
    |__ To here (same place)
    Add a "\n"
}
```

This will result in

```go
func fn() {
    "\n"stmt.List[0]
}
```

Which is

```go
func fn() {
    
stmt.List[0]
}
```

And since applying suggestions with the `-fix` flag also formats the code according to `go fmt`, the result will be

```go
func fn() {
    
	stmt.List[0]
}
```

Reply:
Great explanation, thanks for taking the time